pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -895,14 +1012,198 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+				return
+			}
+		}
+	}
+
+	// Step 8: start containers in podContainerChanges.ContainersToStart.
 	for _, idx := range podContainerChanges.ContainersToStart {
 		start("container", containerStartSpec(&pod.Spec.Containers[idx]))
 	}
 
 	return
 }
 
+func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, podStatus *kubecontainer.PodStatus, resourceName string, containersToUpdate []containerToUpdateInfo) error {
+	type updatedResourcesInfo struct {
+		cStatus *kubecontainer.Status
+		rLimits v1.ResourceList
+		rAllocs v1.ResourceList
+	}
+	updatedResourcesMap := make(map[kubecontainer.ContainerID]updatedResourcesInfo)
+
+	for _, cInfo := range containersToUpdate {
+		container := cInfo.apiContainer
+		containerStatus := cInfo.apiContainerStatus
+		specLimits := container.Resources.Limits
+		statusAllocs := containerStatus.ResourcesAllocated
+		defer func() {
+			container.Resources.Limits = specLimits
+			containerStatus.ResourcesAllocated = statusAllocs
+		}()
+		statusLimits := containerStatus.Resources.Limits
+		statusRequests := containerStatus.Resources.Requests
+		// Runtime container status resources, if set, takes precedence.
+		if cInfo.kubeContainerStatus.Resources.Limits != nil {
+			statusLimits = cInfo.kubeContainerStatus.Resources.Limits
+		}
+		if cInfo.kubeContainerStatus.Resources.Requests != nil {
+			statusRequests = cInfo.kubeContainerStatus.Resources.Requests
+		}
+		switch resourceName {
+		case cpuRequest:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *statusLimits.Cpu(),
+				v1.ResourceMemory: *statusLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusAllocs.Cpu(),
+				v1.ResourceMemory: *statusRequests.Memory(),
+			}
+		case cpuLimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *specLimits.Cpu(),
+				v1.ResourceMemory: *statusLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusRequests.Cpu(),
+				v1.ResourceMemory: *statusRequests.Memory(),
+			}
+		case memLimit:
+			container.Resources.Limits = v1.ResourceList{
+				v1.ResourceCPU:    *statusLimits.Cpu(),
+				v1.ResourceMemory: *specLimits.Memory(),
+			}
+			containerStatus.ResourcesAllocated = v1.ResourceList{
+				v1.ResourceCPU:    *statusRequests.Cpu(),
+				v1.ResourceMemory: *statusAllocs.Memory(),
+			}
+		}
+		if err := m.updateContainerResources(pod, container, cInfo.kubeContainerStatus.ID); err != nil {
+			klog.ErrorS(err, "updateContainerResources failed", "container", container.Name, "cID", cInfo.kubeContainerStatus.ID, "pod", format.Pod(pod))
+			return err
+		}
+		cInfo.apiContainerStatus.Resources.Limits = container.Resources.Limits
+		// NOTE: Ideally, cpu and memory resources information come from ContainerStatus CRI API query.
+		//       However, the runtime may not have implemented query support for currently configured resources.
```

Comment:
If runtime has not implemented any of them, runtime should return an error instead.

I don't feel like we should do the hack here to cover that.

Reply:
This is not a hack. This is an new extension to ContainerStatus, a new field they know nothing about. We need to give the runtime folks time to catch up to this .. educate runtime folks to implement this during KubeCon sessions, conformance tests warnings, and eventually failing them if they don't support. I imagine it might take a few release cycles. 

We can certainly live without this by assuming that if UpdateContainerResources CRI didn't fail, the desired settings were applied. But it is better to get the actual config from the horse's mouth.

Reply:
What runtime do you mean here? containerd and cri-o?

I think we should define the CRI behavior first, and implement it in the container runtime, and then support the feature in Kubernetes. If you use a container runtime that doesn't fully support the this function, it should surface an error, instead of we trying to help them generate container status in CRI.

Reply:
One thing we can probably do, is to call `ContainerStatus` and check whether the status fields are updated after a successful update. If not, return an error explicitly.

I feel like that seems safer than trying to update the status here. I'd rather we fail explicitly for this bad setup, instead of complicating the kubelet logic even more by bypassing pod cache to generate container status.

Reply:
> One thing we can probably do, is to call `ContainerStatus` and check whether the status fields are updated after a successful update. If not, return an error explicitly.
> 

I agree, that sounds reasonable. The current code queries ContainerStatus (GetPodStatus) but if ContainerStatus.Resources is empty, it assumes all is well if UpdateContainerResources is successful, and this hides problems. I'll fail it instead.

Fortunately with CRI, it is a small community and that makes it easy to influence and get things done. We already have dockershim covered here. We stay in alpha until containerd, cri-o, and maybe kata releases have this implemented. Maybe we can drive PRs with them too. Getting more runtimes is a bonus.

Reply:
I just remembered something from an earlier discussion in one of our SIG-node meetings - if the ContainerStatus does not turnaround and give us the desired config immediately after UpdateContainerResouces, we cannot just fault it right-away in the resize codepath. Runtimes (especially VM runtimes) may take some time to actuate the new configuration. We need to give some slack.

We discussed the possibility of failing kubelet on container create where resources are not reported in ContainerStatus for pods that configure resources. Alternately we can leverage conformance tests to qualify runtimes.

Reply:
Got rid of the hiding part and works fine with dockershim.

Reply:
> I just remembered something from an earlier discussion in one of our SIG-node meetings - if the ContainerStatus does not turnaround and give us the desired config immediately after UpdateContainerResouces, we cannot just fault it right-away in the resize codepath. Runtimes (especially VM runtimes) may take some time to actuate the new configuration. We need to give some slack.

If that is the case, we can probably just **check empty resource fields** in `ContainerStatus`, which indicates the container runtime doesn't support it. We should define that clearly in CRI.

Reply:
There are two cases here:
1. Runtime does not support ContainerStatus.Resources and this field is empty for containers that have Resources. 
1. Runtime supports it but the values returned by ContainerStatus() immediately following UpdateContainerResources() may not be what you expect - runtime is actuating and may take time.

In case 1, we fail kubelet at first sign of trouble. In case 2, we log error and keep retrying (to a certain point). 

I'm tracking all outstanding issues, including this, to knock out before beta on the main enhancement tracking issue. https://github.com/kubernetes/enhancements/issues/1287

Reply:
How is this handled at the end? I can't find the corresponding code.

Reply:
I hope the code re-do addresses mostly this issue. 

To summarize my thoughts (just to see if we are on the same page), we have the following points:
- We call UpdateContainerResources CRI API to change resources on a container.
  - If runtime returns success, we assume that the new resources were accepted (they may or may not be effective upon CRI return)
  - If runtime returns error, we bail at that very moment (no further downstream container updates are invoked) and retry later.
- We call GetPodStatus (ContainerStatus CRI API) to query resources when resize is in progress.
  - During alpha, we push runtime folks to handle the newly Resources field added in ContainerStatus CRI API.
    - Socializing the change to runtime groups and devs, pointing to dockershim as reference/example.
    - Bring it up during the next KubeCon
  - Post alpha, if runtime returns empty resources for a container that has resources specified, we halt kubelet.